// @http2/configuration/schema/host.js
module.exports.hostSchema = {
  id: '/Host',
  type: 'object',
  additionalProperties: false,
  properties: {
    domain: {
      anyOf: [
        {type: 'string', format: 'idn-hostname'},
        {type: 'string', format: 'ipv4'},
        {type: 'string', format: 'ipv6'}
      ]
    },
    root: {type: 'string'},
    fallback: {
      oneOf: [
        {type: 'string'},
        {
          type: 'object',
          additionalProperties: false,
          patternProperties: {
            '[245]\\d\\d': {type: 'string'}
          }
        }
      ]
    },
    cacheControl: {
      type: 'object',
      additionalProperties: false,
      properties: {
        immutable: {
          type: 'array',
          uniqueItems: true,
          items: {
            type: 'string',
            anyOf: [
              {enum: ['emoji', 'hex']},
              {format: 'extglob'}
            ]
          }
        }
      }
    },
    directories: {
      type: 'object',
      additionalProperties: false,
      properties: {
        trailingSlash: {
          type: 'string',
          enum: ['always', 'never']
        }
      }
    },
    accessControl: {
      type: 'object',
      additionalProperties: false,
      properties: {
        allowOrigin: {
          anyOf: [
            {type: 'string', enum: ['*']},
            {type: 'string', format: 'uri'}
          ]
        }
      }
    },
    serviceWorker: {
      type: 'object',
      additionalProperties: false,
      properties: {
        allowed: {type: 'string'}
      }
    },
    strictTransportSecurity: {
      type: 'object',
      properties: {
        maxAge: {type: 'integer', minimum: 0}
      }
    },
    manifest: {'$ref': '/Manifest'}
  }
}
