// @http2/configuration/schema/options.js
module.exports.optionsSchema = {
  id: '/Options',
  type: 'object',
  definitions: {
    port: {
      type: 'integer',
      minimum: 1,
      maximum: 2 ** 16 - 1
    }
  },
  additionalProperties: false,
  properties: {
    signature: {type: 'boolean'},
    acme: {
      type: 'object',
      additionalProperties: false,
      properties: {
        proxy: {
          anyOf: [
            {type: 'string', format: 'uri'},
            {type: 'string', value: ''}
          ]
        },
        redirect: {
          anyOf: [
            {type: 'string', format: 'uri'},
            {type: 'string', value: ''}
          ]
        },
        store: {type: 'string'},
        key: {type: 'string'},
        cert: {type: 'string'},
        webroot: {type: 'string'}
      }
    },
    log: {
      type: 'object',
      additionalProperties: false,
      properties: {
        level: {
          type: 'string',
          enum: ['silent', 'fatal', 'error', 'warn', 'info', 'debug', 'trace']
        }
      }
    },
    http: {
      type: 'object',
      additionalProperties: false,
      properties: {
        redirect: {type: 'boolean'},
        from: {'$ref': '#/definitions/port'},
        to: {'$ref': '#/definitions/port'}
      }
    },
    https: {
      type: 'object',
      additionalProperties: false,
      properties: {
        key: {type: 'string'},
        cert: {type: 'string'},
        ca: {
          type: 'array',
          uniqueItems: true,
          items: {type: 'string'}
        },
        port: {'$ref': '#/definitions/port'}
      }
    },
    hosts: {
      type: 'array',
      items: {'$ref': '/Host'}
    },
    workers: {
      type: 'object',
      additionalProperties: false,
      properties: {
        count: {
          anyOf: [
            {type: 'integer', minimum: 1},
            {type: 'string', format: 'calculation'}
          ]
        }
      }
    }
  }
}
